snapcraft/parts/plugins/kernel.py

```diff
@@ -0,0 +1,1750 @@
+
+        if values.get("kernel_initrd_compression_options") and not values.get(
+            "kernel_initrd_compression"
+        ):
+            raise ValueError(
+                "kernel-initrd-compression-options requires also kernel-initrd-compression to be defined."
+            )
+        return values
+
+    @classmethod
+    def unmarshal(cls, data: Dict[str, Any]):
+        """Populate class attributes from the part specification.
+
+        :param data: A dictionary containing part properties.
+
+        :return: The populated plugin properties data object.
+
+        :raise pydantic.ValidationError: If validation fails.
+        """
+        plugin_data = plugins.extract_plugin_properties(
+            data, plugin_name="kernel", required=[]
+        )
+        return cls(**plugin_data)
+
+
+class KernelPlugin(plugins.Plugin):
+    """Plugin for the kernel snap build."""
+
+    properties_class = KernelPluginProperties
+
+    def __init__(
+        self, *, properties: plugins.PluginProperties, part_info: infos.PartInfo
+    ) -> None:
+        super().__init__(properties=properties, part_info=part_info)
+        self.options = cast(KernelPluginProperties, self._options)
+        self._get_deb_architecture()
+        self._get_kernel_architecture()
+        # check if we are cross building
+        self._cross_building = False
+        if (
+            self._part_info.project_info.host_arch
+            != self._part_info.project_info.target_arch
+        ):
+            self._cross_building = True
+        self._llvm_version = self._determine_llvm_version()
+
+    def _determine_llvm_version(self) -> Optional[str]:
+        if (
+            isinstance(self.options.kernel_use_llvm, bool)
+            and self.options.kernel_use_llvm
+        ):
+            return "1"
+        if isinstance(self.options.kernel_use_llvm, str):
+            suffix = re.match(r"^-\d+$", self.options.kernel_use_llvm)
+            if suffix is None:
+                raise ValueError(
+                    f'kernel-use-llvm must match the format "-<version>" (e.g. "-12"), not "{self.options.kernel_use_llvm}"'
+                )
+            return self.options.kernel_use_llvm
+        # Not use LLVM utilities
+        return None
+
+    def _init_build_env(self) -> None:
+        # first get all the architectures, new v2 plugin is making life difficult
+        logger.info("Initializing build env...")
+
+        self.make_cmd = ["make", "-j$(nproc)"]
+        # we are building out of tree, configure paths
+        self.make_cmd.append("-C")
+        self.make_cmd.append("${KERNEL_SRC}")
+        self.make_cmd.append("O=${CRAFT_PART_BUILD}")
+
+        self._check_cross_compilation()
+        self._set_kernel_targets()
+        self._set_llvm()
+
+        # determine type of initrd
+        snapd_snap_file_name = _SNAPD_SNAP_FILE.format(
+            snap_name=_SNAPD_SNAP_NAME,
+            architecture=self._part_info.project_info.target_arch,
```

Comment:
We can have a `target_arch` attribute set from the part info to match the core20 plugin.